### PR TITLE
Initial mbedTLS 3.0.0 support

### PR DESCRIPTION
**Pull request by Benau**
Tested in android and windows and works

Notice: I'm not sure if I've done the tls handling correctly as newest mbedtls 3 drops tls 1.1, 1.2 supports, comments welcome....

And I've not heard from mbedtls developer about prettier access to private member of some structures, so I have to use MBEDTLS_PRIVATE macros...

see https://github.com/curl/curl/discussions/7376#discussioncomment-995479

### CODE REVIEW

**Comment by bagder** [comment]
It seems this breaks [the build of the older mbedtls version](https://curl.zuul.vexxhost.dev/build/94e2171f6b4d41fea5a4ac83667819f8)?

**Comment by Benau** [comment]
ok sha256.c has an inconsistent ```HAS_RESULT_CODE_BASED_FUNCTIONS``` definition at the top in the past vs ```HAS_MBEDTLS_RESULT_CODE_BASED_FUNCTIONS``` (rest of the files), I assume you don't mind me fix that there....

File: lib/md4.c [hunk 1]
```diff
@@ -36,12 +36,17 @@
 #endif /* USE_OPENSSL */
 
 #ifdef USE_MBEDTLS
-#include <mbedtls/config.h>
 #include <mbedtls/version.h>
+#if MBEDTLS_VERSION_MAJOR < 3
```
Comment (danielgustafsson) [inline lib/md4.c:40]: Since the change that altered the header includes was introduced in 3, the check should be reversed and include the new file iff we are compiling against 3 and not the other way around. Nitpick to some extent, but it will improve readability.

Also, why doesn't this (and the remaining changes) use `MBEDTLS_VERSION_NUMBER` as the rest if the code does?

File: lib/vtls/mbedtls.c [hunk 2]
```diff
@@ -647,14 +681,25 @@ mbed_connect_step2(struct Curl_easy *data, struct connectdata *conn,
     /* Make a copy of our const peercert because mbedtls_pk_write_pubkey_der
        needs a non-const key, for now.
        https://github.com/ARMmbed/mbedtls/issues/396 */
+#if MBEDTLS_VERSION_MAJOR >= 3
+    if(mbedtls_x509_crt_parse_der(p,
+                        peercert->MBEDTLS_PRIVATE(raw).MBEDTLS_PRIVATE(p),
+                        peercert->MBEDTLS_PRIVATE(raw).MBEDTLS_PRIVATE(len))) {
+#else
```
Comment (danielgustafsson) [inline lib/vtls/mbedtls.c:688]: This might be a dumb question, but I don't know mbedtls at all. Does this mean that the fields we need are now private but were public in the past, or have they always been private?

File: lib/vtls/mbedtls.c [hunk 3]
```diff
@@ -184,12 +186,17 @@ static Curl_send mbed_send;
 static CURLcode mbedtls_version_from_curl(int *mbedver, long version)
 {
   switch(version) {
+#if MBEDTLS_VERSION_MAJOR >= 3
+    case CURL_SSLVERSION_TLSv1_0:
+    case CURL_SSLVERSION_TLSv1_1:
+#else
```
Comment (danielgustafsson) [inline lib/vtls/mbedtls.c:192]: This makes the code too hard to parse. Since the TLS support is so vastly different, I think we should provide two different implementations of `mbedtls_version_from_curl` entirely and do the version preprocessing around those rather than pick this one apart.

File: lib/vtls/mbedtls.c [hunk 4]
```diff
@@ -1069,7 +1114,11 @@ static CURLcode mbedtls_sha256sum(const unsigned char *input,
   mbedtls_sha256(input, inputlen, sha256sum, 0);
 #else
   /* returns 0 on success, otherwise failure */
+#if MBEDTLS_VERSION_MAJOR >= 3
+  if(mbedtls_sha256(input, inputlen, sha256sum, 0) != 0)
+#else
   if(mbedtls_sha256_ret(input, inputlen, sha256sum, 0) != 0)
+#endif
```
Comment (danielgustafsson) [inline lib/vtls/mbedtls.c:1121]: The fact that we now use the same function name in 3 that we used in 2.0.x which was deprecated in favour of `_ret` in 2.x needs a pretty big comment explaining how that mess is strung together.

Also, while not a fault of this PR, blimey thats ugly..

**Review by danielgustafsson — COMMENTED** [review]
I would not be comfortable in merging this unless we can an mbedTLS 3.0 job in the CI, are there stable packages around to make that happen?

**Comment by bagder** [comment]
> I would not be comfortable in merging this unless we can an mbedTLS 3.0 job in the CI, are there stable packages around to make that happen?

I presume we can always get it from git and build our own version. But yes, I too would like to have that in the CI when we merge v3 support.

File: lib/vtls/mbedtls.c (hunk 2 above)
Comment (Benau) [reply lib/vtls/mbedtls.c:696]: quote from mbedtls developer:
`
Fields in structures are now private by default. We're aware that we may have been overeager in some places, and we're likely to make a few fields public again in 3.x minor releases, or to add new accessor functions.
`
https://github.com/curl/curl/discussions/7376#discussioncomment-993419

**Comment by Benau** [comment]
I made the requested changes now by the way

except i'm not a good documentation writter for "a pretty big comment explaining how that mess is strung together"...

File: lib/sha256.c [hunk 5]
```diff
@@ -43,7 +43,7 @@
 #include <mbedtls/version.h>
 
 #if(MBEDTLS_VERSION_NUMBER >= 0x02070000)
-  #define HAS_RESULT_CODE_BASED_FUNCTIONS
+  #define HAS_MBEDTLS_RESULT_CODE_BASED_FUNCTIONS
```
Comment (bagder) [inline lib/sha256.c:46]: Wouldn't the code be easier if this code block instead was made like?
~~~c
#if(MBEDTLS_VERSION_NUMBER >= 0x02070000) && \
   (MBEDTLS_VERSION_NUMBER < 0x03000000)
  #define HAS_MBEDTLS_RESULT_CODE_BASED_FUNCTIONS
#endif

**Comment by danielgustafsson** [comment]
> except i'm not a good documentation writter for "a pretty big comment explaining how that mess is strung together"...

That's fine, we're all in this together and can collaborate. I would recommend adding a `TODO:` comment which states that this would need to get done, so that we don't forget before it goes in.

File: lib/vtls/mbedtls.c (hunk 2 above)
Comment (danielgustafsson) [reply lib/vtls/mbedtls.c:696]: That sounds really scary, the fact that a member is private means that it can go away or radically change at any time. Relying on that not happening seems pretty terrible, but that's again not the fault of this patch.

**Comment by Benau** [comment]
done

File: lib/md4.c [hunk 6]
```diff
@@ -36,12 +36,17 @@
...
+#if MBEDTLS_VERSION_NUMBER >= 0x03000000
+#include <mbedtls/mbedtls_config.h>
+#else
+#include <mbedtls/config.h>
+#endif
 
 #if(MBEDTLS_VERSION_NUMBER >= 0x02070000)
   #define HAS_MBEDTLS_RESULT_CODE_BASED_FUNCTIONS
 #endif
+
```
Comment (jsoref) [inline lib/md4.c:49]: This blank line seems odd... (compare lin 39/40 of lib/md5.c below)

File: lib/vtls/mbedtls.c [hunk 7]
```diff
@@ -1064,12 +1117,17 @@ static CURLcode mbedtls_sha256sum(const unsigned char *input,
                                   unsigned char *sha256sum,
                                   size_t sha256len UNUSED_PARAM)
 {
+  /* TODO: explain this for different mbedtls 2.x vs 3 version */
   (void)sha256len;
 #if MBEDTLS_VERSION_NUMBER < 0x02070000
   mbedtls_sha256(input, inputlen, sha256sum, 0);
 #else
   /* returns 0 on success, otherwise failure */
+#if MBEDTLS_VERSION_NUMBER >= 0x03000000
```
Comment (jsoref) [inline lib/vtls/mbedtls.c:1126]: In general, wouldn't the preference be to put this before the MBEDTLS_VERSION_NUMBER < 0x02070000?

File: lib/md4.c [hunk 8]
```diff
@@ -36,12 +36,17 @@
 #endif /* USE_OPENSSL */
 
 #ifdef USE_MBEDTLS
```
Comment (jsoref) [inline lib/md4.c:38]: note to self: this is an `#ifdef`

File: lib/sha256.c [hunk 9]
```diff
@@ -105,8 +106,8 @@ typedef mbedtls_sha256_context SHA256_CTX;
 
 static void SHA256_Init(SHA256_CTX *ctx)
 {
-#if !defined(HAS_RESULT_CODE_BASED_FUNCTIONS)
-  mbedtls_sha256_starts(ctx, 0);
+#if !defined(HAS_MBEDTLS_RESULT_CODE_BASED_FUNCTIONS)
```
Comment (jsoref) [inline lib/sha256.c:109]: Is there a style guide for `#if !defined...` vs `#ifndef`?

https://github.com/curl/curl/search?q=ifndef it looks like both flavors are used...

offhand, it looks like `#if !defined...` tends to be used with a boolean operator for when one is performing multiple checks concurrently.
https://github.com/curl/curl/search?p=2&q=%22%23if+%21defined%22

(Yes, I understand this is preexisting code.)

**Review by jsoref — COMMENTED** [review]
Some drive-by-thoughts. Feel free to ignore.

I'd be happy to offer comments on the TODO text, but it doesn't make enough sense to me to comment at this time.

File: lib/sha256.c (hunk 9 above)
Comment (bagder) [reply lib/sha256.c:109]: > Is there a style guide for #if !defined... vs #ifndef?

We prefer #ifdef and #ifndef if you only check a single define. If you check more than one, then #if needs to be used.

**Comment by bagder** [comment]
Can you please squash the commits, rebase and force-push to make this easier to review? Then I figure we only need a CI job added for mbedTLS. I might take a stab at that.

**Comment by Benau** [comment]
Done

**Comment by wyattoday** [comment]
Any chance of getting this merged soon? This fixes https://github.com/curl/curl/issues/7385, and looks about how we solved most of the problems before we ran out of time.

**Comment by bagder** [comment]
Once we have a CI build setup that verifies this build I'll be ready to merge.

**Comment by bagder** [comment]
Thanks!
